File: diedie/src/diedie_ffi.erl
-module(diedie_ffi).

-export([exit/2]).

exit(Code, MaybeMessage) when is_integer(Code) ->
    case {Code, MaybeMessage} of
        {0, {some, Message}} ->
            io:format("Quit: ~p~n", [Message]);
        {Code, {some, Message}} ->
            io:format("Abort (~p): ~p~n", [Code, Message]);
        {Code, none} ->
            io:format("Abort (~p)~n", [Code])
    end,
    erlang:halt(Code).

File: diedie/src/diedie_ffi.mjs
export function exit(code, maybe_message) {
  let message = (function () {
    switch (maybe_message.constructor.name) {
      case "Some":
        return maybe_message["0"];
      case "None":
        return null;
    }
  })();

  if (code === 0 && message !== null) {
    console.log(`Quit: ${message}\n`);
  } else if (code !== 0 && message !== null) {
    console.log(`Abort (${code}): ${message}\n`);
  } else if (code !== 0) {
    console.log(`Abort (${code})\n`);
  }

  process.exit(code);
}
